File: apps/preview/src/project_edit.rs
//! Typed document operations; these never mutate the native render graph directly.
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct MixValues {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pan: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mute: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub solo: Option<bool>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ProjectEdit {
    EffectOrder {
        owner: String,
        index: usize,
        order: Vec<usize>,
    },
    Channel {
        index: usize,
        values: MixValues,
    },
    Bus {
        index: usize,
        values: MixValues,
    },
    Track {
        index: usize,
        #[serde(skip_serializing_if = "Option::is_none")]
        enabled: Option<bool>,
        #[serde(skip_serializing_if = "Option::is_none")]
        mute: Option<bool>,
        #[serde(skip_serializing_if = "Option::is_none")]
        solo: Option<bool>,
    },
    Tempo {
        bpm: f64,
    },
    Instrument {
        index: usize,
        config: serde_json::Value,
    },
    Effect {
        owner: String,
        index: usize,
        #[serde(skip_serializing_if = "Option::is_none")]
        slot: Option<usize>,
        #[serde(skip_serializing_if = "Option::is_none")]
        config: Option<serde_json::Value>,
    },
}

File: packages/protocol/src/document/project-edit.ts
import { z } from "zod";
import { effectRefSchema, instrumentRefSchema } from "../authoring/refs.js";

const index = z.number().int().nonnegative();
const mix = z
  .object({
    level: z.number().finite().min(0).max(2).optional(),
    pan: z.number().finite().min(-1).max(1).optional(),
    mute: z.boolean().optional(),
    solo: z.boolean().optional(),
  })
  .strict();
/** Source-level operations address builder order; execution IDs never enter authored code. */
export const projectEditSchema = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("channel"), index, values: mix }).strict(),
  z.object({ kind: z.literal("bus"), index, values: mix }).strict(),
  z
    .object({
      kind: z.literal("track"),
      index,
      enabled: z.boolean().optional(),
      mute: z.boolean().optional(),
      solo: z.boolean().optional(),
    })
    .strict()
    .refine(
      (edit) => edit.enabled !== undefined || edit.mute !== undefined || edit.solo !== undefined,
      "track edit is empty",
    ),
  z.object({ kind: z.literal("tempo"), bpm: z.number().finite().min(20).max(999) }).strict(),
  z
    .object({
      kind: z.literal("effectOrder"),
      owner: z.enum(["channel", "bus"]),
      index,
      order: z.array(index).max(256),
    })
    .strict(),
  z
    .object({ kind: z.literal("instrument"), index, config: instrumentRefSchema.omit({ instanceId: true }).strict() })
    .strict(),
  z
    .object({
      kind: z.literal("effect"),
      owner: z.enum(["channel", "bus"]),
      index,
      slot: index.optional(),
      config: effectRefSchema.omit({ instanceId: true }).strict().optional(),
    })
    .strict(),
]);
export type ProjectEdit = z.infer<typeof projectEditSchema>;
